tower-http/src/request_id.rs

```diff
@@ -0,0 +1,575 @@
+    }
+}
+
+impl<S> Layer<S> for PropagateRequestIdLayer {
+    type Service = PropagateRequestId<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        PropagateRequestId::new(inner, self.header_name.clone())
+    }
+}
+
+/// Propagate request ids from requests to responses.
+///
+/// See the [module docs](self) for an example.
+///
+/// If the request contains a matching header that header will be applied to responses. If a
+/// [`RequestId`] extension is also present it will be propagated as well.
+#[derive(Debug, Clone)]
+pub struct PropagateRequestId<S> {
+    inner: S,
+    header_name: HeaderName,
+}
+
+impl<S> PropagateRequestId<S> {
+    /// Create a new `PropagateRequestId`.
+    pub fn new(inner: S, header_name: HeaderName) -> Self {
+        Self { inner, header_name }
+    }
+
+    /// Create a new `PropagateRequestId` that uses `x-request-id` as the header name.
+    pub fn x_request_id(inner: S) -> Self {
+        Self::new(inner, HeaderName::from_static(X_REQUEST_ID))
+    }
+
+    define_inner_service_accessors!();
+
+    /// Returns a new [`Layer`] that wraps services with a `PropagateRequestId` middleware.
+    pub fn layer(header_name: HeaderName) -> PropagateRequestIdLayer {
+        PropagateRequestIdLayer::new(header_name)
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for PropagateRequestId<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = PropagateRequestIdResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let request_id = req
+            .headers()
+            .get(&self.header_name)
+            .cloned()
+            .map(RequestId::new);
+
+        PropagateRequestIdResponseFuture {
+            inner: self.inner.call(req),
+            header_name: self.header_name.clone(),
+            request_id,
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`PropagateRequestId`].
+    pub struct PropagateRequestIdResponseFuture<F> {
+        #[pin]
+        inner: F,
+        header_name: HeaderName,
+        request_id: Option<RequestId>,
+    }
+}
+
+impl<F, B, E> Future for PropagateRequestIdResponseFuture<F>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+{
+    type Output = Result<Response<B>, E>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let mut response = futures_core::ready!(this.inner.poll(cx))?;
+
+        if let Some(current_id) = response.headers().get(&*this.header_name) {
+            if response.extensions().get::<RequestId>().is_none() {
+                let current_id = current_id.clone();
+                response.extensions_mut().insert(current_id);
```

Comment:
I don't think putting it inside `Response`'s extensions is useful?
Generally you just want to return back header to user for request identification
If you're concerned about using logging middleware, then it is not necessary too.
In my case I just extract request id from request and pass it to future alongside future that resolves response

Reply:
I think its useful to also have available in response extensions as it makes it easier to access the request id if you don't know which header value is being used.

Reply:
Maybe, but I just do not see case where you can have access to response but not request itself?

Reply:
In a `.map_response` combinator.

Reply:
Ah I see if there is use case then it makes sense